### ISSUE TITLE
Random performance on other dataset, Why it Does not learns?

### ISSUE THREAD

**Issue by rezazad68** [XifengGuo/CapsNet-Keras#48]
I applied your implementation on 2 other datasets. However, it works like a random classification with very low performance. Firstly I trained it on Dog vs Cat (kaggle problem) it can't learns this binary classification and has accuracy around 52 after 50 epochs, While the simple CNN can learns it very well. I also applied it on my own dataset with 20 classes but its accuracy was 6 percentage! Could please let me know what is the main reason of this important issue? Thanks

**Comment by XifengGuo** [comment]
@rezazad68 Did the training loss decrease? If it didn't, I assume the initial value of W in CapsuleLayer is bad, which may make the output of this layer smaller than 1e-8. Try to initialize W with random value in [-1, 1].

**Comment by rezazad68** [comment]
I changed the following initializers with the 'TruncatedNormal', 'lecun_uniform', 'glorot_normal', 'glorot_uniform', non of them worked. It still has random performance without decreasing the loss value. could you please let me know how to fix this. thanks

    def __init__(self, num_capsule, dim_capsule, routings=3,
                 kernel_initializer='**TruncatedNorma**l',
                 **kwargs)

**Comment by XifengGuo** [comment]
Try this. I don't know if it works.   
```
from keras import initializers
```
```    
def __init__(self, num_capsule, dim_capsule, routings=3,
                 kernel_initializer='random_uniform',
                 **kwargs):
        super(CapsuleLayer, self).__init__(**kwargs)
        self.num_capsule = num_capsule
        self.dim_capsule = dim_capsule
        self.routings = routings
        self.kernel_initializer = initializers.random_uniform(-1, 1)
```

**Comment by shivamDuggal** [comment]
Hi,
I applied the same structure model on another dataset. The loss doesn't seem to converge. I even tried training only on 2 images, and the behaviour observed is same. The gradients for all the layers is almost zero, even the last capsule layer. I have attached the tensorboard histograms for gradients and the weights.
<img width="1269" alt="tensorboard" src="https://user-images.githubusercontent.com/29490408/35053201-779e98a6-fbcf-11e7-9047-adb98753bff3.png">

**Comment by marhetz** [comment]
Hi, thanks for your implementation, im experiencing the same issue. Im trying binary classification with my own dataset of 4000 imgs/class and the capsnet loss is hardly decreasing and the accuracy stays around 50%. Only Encoder loss is dropping
I tried another weight initialization, but it doesnt change anything. Its working on mnist and cifar10(to an extend) though. 

Does anyone have an idea?

**Comment by mbenami** [comment]
Hi.
I try your code on my test to see if it learn something 
for the first test, I use only 10 class of RGB 128x128 
I add (2 conv + max pooling ) * 2, to reduce the input shape to be close to 28x28 as in the MNIST dataset. 
I got 90% after 2 epoch 

but when I try the same architecture on 270 class it's seemed to not learning at all 
have you tried this architecture on large data-set (more than 10 ) with good results?

**Comment by rezazad68** [comment]
All right then fellas, I tested an experiment on 2 class classification with this model! I simply chose only classes with label 0 and 1 of Mnist dataset and trained this model on this 2 class dataset, however, the capsule net loss didn't change over the training epochs.  only the decoder loss will be decreased. I was wondering to know why this issue happens? is it related to the model structure or initial weights?

**Comment by mbenami** [comment]
Update.
I got good results when I use some pre-trained weights. 
In my experience, I add the first 2 block of vgg16 as input (with pre-trained wights )
and the model starts to converge immediately.

**Comment by FlyDogFan** [comment]
@mbenami  can you show me the detail of how to "add the 1st 2 blocks of vgg16 as input (with Pretrained wights)"

**Comment by mbenami** [comment]
Hi @FlyDogFan 
I change the `CapsNet` function in `capsulenet.py` file to fit my case 

```
from keras.applications import vgg16
def CapsNet(input_shape, n_class, routings):
    """
    A Capsule Network on MNIST.
    :param input_shape: data shape, 3d, [width, height, channels]
    :param n_class: number of classes
    :param routings: number of routing iterations
    :return: Two Keras Models, the first one used for training, and the second one for evaluation.
            `eval_model` can also be used for training.
    """
    # x = layers.Input(shape=input_shape)

    x = vgg16.VGG16(input_shape=input_shape, include_top=False)

    # Layer 1: Just a conventional Conv2D layer
    conv1 = layers.Conv2D(filters=256, kernel_size=9, strides=1, padding='valid', activation='relu', name='conv1')(x.layers[6].output)

    # Layer 2: Conv2D layer with `squash` activation, then reshape to [None, num_capsule, dim_capsule]
    primarycaps = PrimaryCap(conv1, dim_capsule=8, n_channels=32, kernel_size=9, strides=2, padding='valid')

    # Layer 3: Capsule layer. Routing algorithm works here.
    digitcaps = CapsuleLayer(num_capsule=n_class, dim_capsule=16, routings=routings,
                             name='digitcaps')(primarycaps)

    # Layer 4: This is an auxiliary layer to replace each capsule with its length. Just to match the true label's shape.
    # If using tensorflow, this will not be necessary. :)
    out_caps = Length(name='capsnet')(digitcaps)

    # Decoder network.
    y = layers.Input(shape=(n_class,))
    masked_by_y = Mask()([digitcaps, y])  # The true label is used to mask the output of capsule layer. For training
    masked = Mask()(digitcaps)  # Mask using the capsule with maximal length. For prediction

    # Shared Decoder model in training and prediction
    decoder = models.Sequential(name='decoder')
    decoder.add(layers.Dense(512, activation='relu', input_dim=16*n_class))
    decoder.add(layers.Dense(1024, activation='relu'))
    decoder.add(layers.Dense(np.prod(input_shape), activation='sigmoid'))
    decoder.add(layers.Reshape(target_shape=input_shape, name='out_recon'))

    # Models for training and evaluation (prediction)
    train_model = models.Model([x.input, y], [out_caps, decoder(masked_by_y)])
    eval_model = models.Model(x.input, [out_caps, decoder(masked)])

    # manipulate model
    noise = layers.Input(shape=(n_class, 16))
    noised_digitcaps = layers.Add()([digitcaps, noise])
    masked_noised_y = Mask()([noised_digitcaps, y])
    manipulate_model = models.Model([x.input, y, noise], decoder(masked_noised_y))
    return train_model, eval_model, manipulate_model
```
The main changes is to use 
`x = vgg16.VGG16(input_shape=input_shape, include_top=False)` as the first layer of the network 
and change the rest of the code to fit that 

I stop my research in this field as it was not giving me better results and the work was very slow 
hope that in the future some progress will be about the speed so more research can be done

**Comment by SoloyT** [comment]
Hi 
Thanks for your implementation.
I want to implement this code on RGB image for binary classification. Which part of code should be changed?

**Comment by KumarLamic** [comment]
https://stackoverflow.com/questions/51203765/is-capsule-network-really-rotationally-invariant-in-practice
